File: core/src/com/unciv/game/utils/CameraStageBaseScreen.java
package com.unciv.game.utils;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Screen;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.graphics.g2d.Batch;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.Stage;
import com.badlogic.gdx.scenes.scene2d.ui.Label;
import com.badlogic.gdx.scenes.scene2d.ui.Skin;
import com.badlogic.gdx.scenes.scene2d.ui.Table;
import com.badlogic.gdx.scenes.scene2d.ui.TextButton;
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;
import com.badlogic.gdx.utils.Align;
import com.badlogic.gdx.utils.viewport.ExtendViewport;
import com.unciv.civinfo.CivilizationInfo;
import com.unciv.game.UnCivGame;
import com.unciv.models.LinqCollection;

import java.util.Collection;
import java.util.Collections;

public class CameraStageBaseScreen implements Screen {

    public UnCivGame game;
    public Stage stage;
    public Skin skin = new Skin(Gdx.files.internal("skin/flat-earth-ui.json"));
    static Batch batch = new SpriteBatch();

    public CameraStageBaseScreen() {
        this.game = UnCivGame.Current;
        stage = new Stage(new ExtendViewport(1000,600
        ),batch);// FitViewport(1000,600)
        Gdx.input.setInputProcessor(stage);
    }


    @Override
    public void show() {    }

    @Override
    public void render(float delta) {
        Gdx.gl.glClearColor(0, 0, 0.2f, 1);
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);

        stage.act();
        stage.draw();;
    }

    @Override
    public void resize(int width, int height) {
        stage.getViewport().update(width,height,true);
    }

    @Override
    public void pause() {    }

    @Override
    public void resume() {    }

    @Override
    public void hide() {    }

    @Override
    public void dispose() {    }

    private LinqCollection<String> tutorialTexts = new LinqCollection<String>();

    public void displayTutorials(String name, LinqCollection<String> texts){
        if(CivilizationInfo.current().tutorial.contains(name)) return;
        CivilizationInfo.current().tutorial.add(name);

        tutorialTexts.addAll(texts);
        if(!isTutorialShowing) displayTutorial();
    }

    boolean isTutorialShowing=false;

    public void displayTutorial(){
        isTutorialShowing=true;
        final Table tutorialTable = new Table().pad(10);
        tutorialTable.background(ImageGetter.getDrawable("skin/tileTableBackground.png")
                .tint(new Color(0x101050cf)));
        Label label = new Label(tutorialTexts.get(0),skin);
        label.setFontScale(1.5f);
        label.setAlignment(Align.center);
        tutorialTexts.remove(0);
        tutorialTable.add(label).pad(10).row();
        TextButton button = new TextButton("Close",skin);
        button.addListener(new ClickListener(){
            @Override
            public void clicked(InputEvent event, float x, float y) {
                tutorialTable.remove();
                if(!tutorialTexts.isEmpty()) displayTutorial();
                else isTutorialShowing=false;
            }
        });
        tutorialTable.add(button).pad(10);
        tutorialTable.pack();
        tutorialTable.setPosition(stage.getWidth()/2-tutorialTable.getWidth()/2,
                stage.getHeight()/2-tutorialTable.getHeight()/2);
        stage.addActor(tutorialTable);
    }

}

File: core/src/com/unciv/ui/utils/CameraStageBaseScreen.kt
package com.unciv.ui.utils

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Screen
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.GL20
import com.badlogic.gdx.graphics.g2d.BitmapFont
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.scenes.scene2d.Stage
import com.badlogic.gdx.scenes.scene2d.ui.*
import com.badlogic.gdx.scenes.scene2d.utils.Drawable
import com.badlogic.gdx.utils.viewport.ExtendViewport
import com.unciv.MainMenuScreen
import com.unciv.UncivGame
import com.unciv.models.Tutorial
import com.unciv.ui.tutorials.TutorialController
import com.unciv.ui.worldscreen.WorldScreen
import com.unciv.ui.worldscreen.mainmenu.OptionsPopup
import kotlin.concurrent.thread

open class CameraStageBaseScreen : Screen {

    var game: UncivGame = UncivGame.Current
    var stage: Stage

    protected val tutorialController by lazy { TutorialController(this) }

    val keyPressDispatcher = KeyPressDispatcher(this.javaClass.simpleName)

    init {
        val resolutions: List<Float> = game.settings.resolution.split("x").map { it.toInt().toFloat() }
        val height = resolutions[1]

        /** The ExtendViewport sets the _minimum_(!) world size - the actual world size will be larger, fitted to screen/window aspect ratio. */
        stage = Stage(ExtendViewport(height, height), SpriteBatch())

        keyPressDispatcher.install(stage) { hasOpenPopups() }
    }

    override fun show() {}

    override fun render(delta: Float) {
        Gdx.gl.glClearColor(0f, 0f, 0.2f, 1f)
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT)

        stage.act()
        stage.draw()
    }

    override fun resize(width: Int, height: Int) {
        stage.viewport.update(width, height, true)
    }

    override fun pause() {}

    override fun resume() {}

    override fun hide() {}

    override fun dispose() {
        keyPressDispatcher.uninstall()
    }

    fun displayTutorial(tutorial: Tutorial, test: (() -> Boolean)? = null) {
        if (!game.settings.showTutorials) return
        if (game.settings.tutorialsShown.contains(tutorial.name)) return
        if (test != null && !test()) return
        tutorialController.showTutorial(tutorial)
    }

    companion object {
        lateinit var skin:Skin
        fun setSkin() {
            Fonts.resetFont()
            skin = Skin().apply {
                add("Nativefont", Fonts.font, BitmapFont::class.java)
                add("RoundedEdgeRectangle", ImageGetter.getRoundedEdgeRectangle(), Drawable::class.java)
                add("Rectangle", ImageGetter.getDrawable(""), Drawable::class.java)
                add("Circle", ImageGetter.getDrawable("OtherIcons/Circle").apply { setMinSize(20f, 20f) }, Drawable::class.java)
                add("Scrollbar", ImageGetter.getDrawable("").apply { setMinSize(10f, 10f) }, Drawable::class.java)
                add("RectangleWithOutline", ImageGetter.getRectangleWithOutline(), Drawable::class.java)
                add("Select-box", ImageGetter.getSelectBox(), Drawable::class.java)
                add("Select-box-pressed", ImageGetter.getSelectBoxPressed(), Drawable::class.java)
                add("Checkbox", ImageGetter.getCheckBox(), Drawable::class.java)
                add("Checkbox-pressed", ImageGetter.getCheckBoxPressed(), Drawable::class.java)
                load(Gdx.files.internal("Skin.json"))
            }
            skin.get(TextButton.TextButtonStyle::class.java).font = Fonts.font.apply { data.setScale(20 / Fonts.ORIGINAL_FONT_SIZE) }
            skin.get(CheckBox.CheckBoxStyle::class.java).font = Fonts.font.apply { data.setScale(20 / Fonts.ORIGINAL_FONT_SIZE) }
            skin.get(CheckBox.CheckBoxStyle::class.java).fontColor = Color.WHITE
            skin.get(Label.LabelStyle::class.java).font = Fonts.font.apply { data.setScale(18 / Fonts.ORIGINAL_FONT_SIZE) }
            skin.get(Label.LabelStyle::class.java).fontColor = Color.WHITE
            skin.get(TextField.TextFieldStyle::class.java).font = Fonts.font.apply { data.setScale(18 / Fonts.ORIGINAL_FONT_SIZE) }
            skin.get(SelectBox.SelectBoxStyle::class.java).font = Fonts.font.apply { data.setScale(20 / Fonts.ORIGINAL_FONT_SIZE) }
            skin.get(SelectBox.SelectBoxStyle::class.java).listStyle.font = Fonts.font.apply { data.setScale(20 / Fonts.ORIGINAL_FONT_SIZE) }
            skin
        }
    }

    fun onBackButtonClicked(action: () -> Unit) {
        keyPressDispatcher[KeyCharAndCode.BACK] = action
    }

    fun isPortrait() = stage.viewport.screenHeight > stage.viewport.screenWidth
    fun isCrampedPortrait() = isPortrait() &&
            game.settings.resolution.split("x").map { it.toInt() }.last() <= 700

    fun openOptionsPopup() {
        val limitOrientationsHelper = game.limitOrientationsHelper
        if (limitOrientationsHelper == null || !game.settings.allowAndroidPortrait || !isCrampedPortrait()) {
            OptionsPopup(this).open(force = true)
            return
        }
        if (!(this is MainMenuScreen || this is WorldScreen)) {
            throw IllegalArgumentException("openOptionsPopup called on wrong derivative class")
        }
        limitOrientationsHelper.allowPortrait(false)
        thread(name="WaitForRotation") {
            var waited = 0
            while (true) {
                val newScreen = (UncivGame.Current.screen as? CameraStageBaseScreen)
                if (waited >= 10000 || newScreen!=null && !newScreen.isPortrait() ) {
                    Gdx.app.postRunnable { OptionsPopup(newScreen ?: this).open(true) }
                    break
                }
                Thread.sleep(200)
                waited += 200
            }
        }
    }
}
